helix-view/src/document.rs

```diff
@@ -1584,25 +1580,67 @@ impl Document {
     /// Intialize/updates the differ for this document with a new base.
     pub fn set_diff_base(&mut self, diff_base: Vec<u8>) {
         if let Ok((diff_base, ..)) = from_reader(&mut diff_base.as_slice(), Some(self.encoding)) {
-            if let Some(differ) = &self.diff_handle {
-                differ.update_diff_base(diff_base);
-                return;
-            }
-            self.diff_handle = Some(DiffHandle::new(diff_base, self.text.clone()))
+            self.set_diff_base_raw(diff_base);
         } else {
             self.diff_handle = None;
         }
     }
 
-    pub fn version_control_head(&self) -> Option<Arc<Box<str>>> {
-        self.version_control_head.as_ref().map(|a| a.load_full())
+    pub fn set_diff_base_raw(&mut self, diff_base: Rope) {
+        if let Some(differ) = &self.diff_handle {
+            differ.update_diff_base(diff_base);
+            return;
+        }
+        self.diff_handle = Some(DiffHandle::new(diff_base, self.text.clone()))
     }
 
-    pub fn set_version_control_head(
+    pub fn update_diff_base(
         &mut self,
-        version_control_head: Option<Arc<ArcSwap<Box<str>>>>,
+        path: &Path,
+        diff_provider: &Git,
+        diff_source: DiffSource,
+        text: Option<Rope>,
     ) {
-        self.version_control_head = version_control_head;
+        match diff_source {
+            DiffSource::Git => match diff_provider.get_diff_base(path) {
+                Ok(diff_base) => self.set_diff_base(diff_base),
+                Err(err) => {
+                    log::info!("{err:#?}");
+                    log::info!("failed to open diff base for for {}", path.display());
+                    self.diff_handle = None;
+                }
+            },
+            DiffSource::File => {
+                if let Some(text) = text {
+                    self.set_diff_base_raw(text);
+                } else {
```

Comment:
sorry I modified the originial review comment but you were too fast.

Instead of making this a parameter you can simply check this from within the document:


```suggestion
                if !self.is_modified() {
                    self.set_diff_base_raw(self.text.clone());
                } else {
```